openff/interchange/_tests/unit_tests/test_models.py

```diff
@@ -110,3 +111,61 @@ def test_reprs():
     assert "blah" in repr(potential_key)
     assert "mult 2" in repr(potential_key)
     assert "bond order 1.111" in repr(potential_key)
+
+
+def test_bondkey_eq_hash():
+    """
+    When __eq__ is true, the hashes must be equal.
+
+    The converse is not required in Python; hash collisions between unequal
+    objects are allowed and will be handled according to __eq__, possibly
+    with a small runtime cost.
+    """
+
+    assert BondKey(atom_indices=(1, 3)) == (1, 3)
+    assert hash(BondKey(atom_indices=(1, 3))) == hash((1, 3))
+    assert BondKey(atom_indices=(1, 3)) != (1, 4)
+    assert BondKey(atom_indices=(1, 3)) != (3, 1)
+    assert BondKey(atom_indices=(1, 3), bond_order=None) == (1, 3)
+    assert hash(BondKey(atom_indices=(1, 3), bond_order=None)) == hash((1, 3))
+    assert BondKey(atom_indices=(1, 3), bond_order=None) != ((1, 3), None)
+    assert BondKey(atom_indices=(1, 3), bond_order=1.5) != (1, 3)
+    assert BondKey(atom_indices=(1, 3), bond_order=1.5) != ((1, 3), None)
+    assert BondKey(atom_indices=(1, 3), bond_order=1.5) != ((1, 3), 1.5)
```

Comment:
I'm pretty sure these two should be equal?

```python
    def __hash__(self) -> int:
        if self.bond_order is None:
            return hash(tuple(self.atom_indices))
        return hash((tuple(self.atom_indices), self.bond_order))

Reply:
I believe this test is correct - Lines 131, 133 and 134 are all extended tuples, which has not been implemented (as we've discussed elsewhere), and in line 133 the bond order is not `None` so the tuple doesn't match.

The second return line in that hash function is the existing behaviour - `BondKey.__eq__` is 

```py
def __eq__(self, other) -> bool:
return super().__eq__(other) or (
    self.bond_order is None and other == self.atom_indices
)
```

Reply:
Comparing hashes in `__eq__` can also run into issues when comparing objects of different types.